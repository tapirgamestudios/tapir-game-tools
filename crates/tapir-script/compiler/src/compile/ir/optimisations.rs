use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
