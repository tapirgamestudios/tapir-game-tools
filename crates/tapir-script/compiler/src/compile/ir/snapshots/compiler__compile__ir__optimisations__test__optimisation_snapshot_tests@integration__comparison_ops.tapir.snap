---
source: crates/tapir-script/compiler/src/compile/ir/optimisations.rs
expression: output
input_file: crates/tapir-script/compiler/src/compile/ir/snapshot_tests/optimisations/integration/comparison_ops.tapir
---
----------- before -------------
=======================
fn @toplevel() -> 

---- block 3 ----
getprop a.24, 0
b.25 = int 6
temp.26 = b.25
temp.27 = a.24
temp.28 = temp.26 > temp.27
conditional_jump temp.28 0 1

---- block 0 ----
temp.29 = int 15
c.30 = black_box(temp.29)
wait
temp.31 = c.30
storeprop temp.31, 0
jmp 2

---- block 1 ----
jmp 2

---- block 2 ----
b.32 = φ(b.25 from 0, b.25 from 1)
a.40 = φ(a.24 from 0, a.24 from 1)
temp.33 = b.32
getprop temp.34, 0
temp.35 = temp.33 < temp.34
conditional_jump temp.35 4 5

---- block 4 ----
temp.36 = int 20
c.37 = black_box(temp.36)
wait
temp.38 = c.37
storeprop temp.38, 0
jmp 6

---- block 5 ----
jmp 6

---- block 6 ----
a.39 = φ(a.40 from 4, a.40 from 5)
temp.41 = a.39
getprop temp.42, 0
temp.43 = temp.41 != temp.42
conditional_jump temp.43 7 8

---- block 7 ----
temp.44 = int 30
c.45 = black_box(temp.44)
wait
temp.46 = c.45
storeprop temp.46, 0
jmp 9

---- block 8 ----
jmp 9

---- block 9 ----
return 

=======================
fn black_box(x.7) -> int

---- block 0 ----
temp.47 = x.7
return temp.47


----------- empty_phi -------------
=======================
fn @toplevel() -> 

---- block 3 ----
getprop a.24, 0
b.25 = int 6
temp.26 = b.25
temp.27 = a.24
temp.28 = temp.26 > temp.27
conditional_jump temp.28 0 1

---- block 0 ----
temp.29 = int 15
c.30 = black_box(temp.29)
wait
temp.31 = c.30
storeprop temp.31, 0
jmp 2

---- block 1 ----
jmp 2

---- block 2 ----
temp.33 = b.25
getprop temp.34, 0
temp.35 = temp.33 < temp.34
conditional_jump temp.35 4 5

---- block 4 ----
temp.36 = int 20
c.37 = black_box(temp.36)
wait
temp.38 = c.37
storeprop temp.38, 0
jmp 6

---- block 5 ----
jmp 6

---- block 6 ----
temp.41 = a.24
getprop temp.42, 0
temp.43 = temp.41 != temp.42
conditional_jump temp.43 7 8

---- block 7 ----
temp.44 = int 30
c.45 = black_box(temp.44)
wait
temp.46 = c.45
storeprop temp.46, 0
jmp 9

---- block 8 ----
jmp 9

---- block 9 ----
return 

=======================
fn black_box(x.7) -> int

---- block 0 ----
temp.47 = x.7
return temp.47


----------- copy_propagation -------------
=======================
fn @toplevel() -> 

---- block 3 ----
getprop a.24, 0
b.25 = int 6
temp.26 = b.25
temp.27 = a.24
temp.28 = b.25 > a.24
conditional_jump temp.28 0 1

---- block 0 ----
temp.29 = int 15
c.30 = black_box(temp.29)
wait
temp.31 = c.30
storeprop c.30, 0
jmp 2

---- block 1 ----
jmp 2

---- block 2 ----
temp.33 = b.25
getprop temp.34, 0
temp.35 = b.25 < temp.34
conditional_jump temp.35 4 5

---- block 4 ----
temp.36 = int 20
c.37 = black_box(temp.36)
wait
temp.38 = c.37
storeprop c.37, 0
jmp 6

---- block 5 ----
jmp 6

---- block 6 ----
temp.41 = a.24
getprop temp.42, 0
temp.43 = a.24 != temp.42
conditional_jump temp.43 7 8

---- block 7 ----
temp.44 = int 30
c.45 = black_box(temp.44)
wait
temp.46 = c.45
storeprop c.45, 0
jmp 9

---- block 8 ----
jmp 9

---- block 9 ----
return 

=======================
fn black_box(x.7) -> int

---- block 0 ----
temp.47 = x.7
return x.7


----------- dead_store_elimination -------------
=======================
fn @toplevel() -> 

---- block 3 ----
getprop a.24, 0
b.25 = int 6
temp.28 = b.25 > a.24
conditional_jump temp.28 0 1

---- block 0 ----
temp.29 = int 15
c.30 = black_box(temp.29)
wait
storeprop c.30, 0
jmp 2

---- block 1 ----
jmp 2

---- block 2 ----
getprop temp.34, 0
temp.35 = b.25 < temp.34
conditional_jump temp.35 4 5

---- block 4 ----
temp.36 = int 20
c.37 = black_box(temp.36)
wait
storeprop c.37, 0
jmp 6

---- block 5 ----
jmp 6

---- block 6 ----
getprop temp.42, 0
temp.43 = a.24 != temp.42
conditional_jump temp.43 7 8

---- block 7 ----
temp.44 = int 30
c.45 = black_box(temp.44)
wait
storeprop c.45, 0
jmp 9

---- block 8 ----
jmp 9

---- block 9 ----
return 

=======================
fn black_box(x.7) -> int

---- block 0 ----
return x.7


----------- empty_block -------------
=======================
fn @toplevel() -> 

---- block 3 ----
getprop a.24, 0
b.25 = int 6
temp.28 = b.25 > a.24
conditional_jump temp.28 0 2

---- block 0 ----
temp.29 = int 15
c.30 = black_box(temp.29)
wait
storeprop c.30, 0
jmp 2

---- block 2 ----
getprop temp.34, 0
temp.35 = b.25 < temp.34
conditional_jump temp.35 4 6

---- block 4 ----
temp.36 = int 20
c.37 = black_box(temp.36)
wait
storeprop c.37, 0
jmp 6

---- block 6 ----
getprop temp.42, 0
temp.43 = a.24 != temp.42
conditional_jump temp.43 7 9

---- block 7 ----
temp.44 = int 30
c.45 = black_box(temp.44)
wait
storeprop c.45, 0
return 

---- block 9 ----
return 

=======================
fn black_box(x.7) -> int

---- block 0 ----
return x.7
